plugins/Compressor/Compressor.cpp

```diff
@@ -0,0 +1,651 @@
+	Effect(&compressor_plugin_descriptor, parent, key),
+	m_compressorControls(this)
+{
+	m_sampleRate = Engine::mixer()->processingSampleRate();
+
+	m_yL[0] = m_yL[1] = COMP_NOISE_FLOOR;
+
+	m_maxLookaheadVal[0] = 0;
+	m_maxLookaheadVal[1] = 0;
+
+	// 200 ms
+	m_crestTimeConst = exp(-1.f / (0.2f * m_sampleRate));
+
+	connect(&m_compressorControls.m_attackModel, SIGNAL(dataChanged()), this, SLOT(calcAttack()));
+	connect(&m_compressorControls.m_releaseModel, SIGNAL(dataChanged()), this, SLOT(calcRelease()));
+	connect(&m_compressorControls.m_holdModel, SIGNAL(dataChanged()), this, SLOT(calcHold()));
+	connect(&m_compressorControls.m_ratioModel, SIGNAL(dataChanged()), this, SLOT(calcRatio()));
+	connect(&m_compressorControls.m_rangeModel, SIGNAL(dataChanged()), this, SLOT(calcRange()));
+	connect(&m_compressorControls.m_rmsModel, SIGNAL(dataChanged()), this, SLOT(resizeRMS()));
+	connect(&m_compressorControls.m_lookaheadLengthModel, SIGNAL(dataChanged()), this, SLOT(calcLookaheadLength()));
+	connect(&m_compressorControls.m_thresholdModel, SIGNAL(dataChanged()), this, SLOT(calcThreshold()));
+	connect(&m_compressorControls.m_kneeModel, SIGNAL(dataChanged()), this, SLOT(calcKnee()));
+	connect(&m_compressorControls.m_outGainModel, SIGNAL(dataChanged()), this, SLOT(calcOutGain()));
+	connect(&m_compressorControls.m_inGainModel, SIGNAL(dataChanged()), this, SLOT(calcInGain()));
+	connect(&m_compressorControls.m_tiltModel, SIGNAL(dataChanged()), this, SLOT(calcTiltCoeffs()));
+	connect(&m_compressorControls.m_tiltFreqModel, SIGNAL(dataChanged()), this, SLOT(calcTiltCoeffs()));
+	connect(&m_compressorControls.m_limiterModel, SIGNAL(dataChanged()), this, SLOT(redrawKnee()));
+	connect(&m_compressorControls.m_mixModel, SIGNAL(dataChanged()), this, SLOT(calcMix()));
+
+	connect(&m_compressorControls.m_autoAttackModel, SIGNAL(dataChanged()), this, SLOT(calcAutoAttack()));
+	connect(&m_compressorControls.m_autoReleaseModel, SIGNAL(dataChanged()), this, SLOT(calcAutoRelease()));
+
+	connect(&m_compressorControls.m_thresholdModel, SIGNAL(dataChanged()), this, SLOT(calcAutoMakeup()));
+	connect(&m_compressorControls.m_ratioModel, SIGNAL(dataChanged()), this, SLOT(calcAutoMakeup()));
+	connect(&m_compressorControls.m_kneeModel, SIGNAL(dataChanged()), this, SLOT(calcAutoMakeup()));
+	connect(&m_compressorControls.m_autoMakeupModel, SIGNAL(dataChanged()), this, SLOT(calcAutoMakeup()));
+
+	connect(Engine::mixer(), SIGNAL(sampleRateChanged()), this, SLOT(changeSampleRate()));
+	changeSampleRate();
+}
+
+
+
+
+CompressorEffect::~CompressorEffect()
+{
+}
+
+
+float CompressorEffect::msToCoeff(float ms)
+{
+	// Convert time in milliseconds to applicable lowpass coefficient
+	return exp(m_coeffPrecalc / ms);
+}
+
+
+
+void CompressorEffect::calcAutoMakeup()
+{
+	// Formulas using the compressor's Threshold, Ratio, and Knee values to estimate a good makeup gain value
+
+	float tempGainResult;
+	if (-m_thresholdVal < -m_kneeVal)// Below knee
```

Comment:
If I read the code correctly, `m_thresholdVal` is in `[-60, 0]` and `m_kneeVal` is in `[0, 48]`(both in dBFS). So I think you're doing something wrong with this conditionals.

Reply:
The problem here is that the tooltips and such say knee is in terms of dBFS rather than dB.  `m_kneeVal` is the distance between the threshold and the edge of the knee (note that variable is half the value of the knob), so the code is entirely correct.  The labels mix up dBFS and dB though, so I'll fix those.

Reply:
Yes, the unit of `m_kneeVal` is dB, indeed. The formula itself is correct, but this block will never execute.

Reply:
One more minor comment: you're using `db` and `dbFs` instead of `dB` and `dBFS`.

Reply:
> One more minor comment: you're using db and dbFs instead of dB and dBFS.

Yeah, just barely fixed that a couple minutes ago when I noticed it.  :+1: 

> Yes, the unit of m_kneeVal is dB, indeed. The formula itself is correct, but this block will never execute.

Ah, you're right.  In an older version of this plugin I allowed the Threshold to go up to +12 dbFs, but that's not longer allowed.